kafka-connect-s3/src/test/java/io/confluent/connect/s3/TopicPartitionWriterTest.java

```diff
@@ -915,78 +923,100 @@ public void testAddingS3ObjectTags() throws Exception{
+    testExceptionReportedToDLQ(faultyRecord, DataException.class, exceptionMessage, true);
   }
 
   @Test
-  public void testExceptionOnNullHeaders() {
+  public void testExceptionOnNullHeadersReported() throws Exception {
     String recordValue = "1";
     int kafkaOffset = 1;
     SinkRecord faultyRecord = new SinkRecord(TOPIC, PARTITION, Schema.STRING_SCHEMA, "key",
         Schema.STRING_SCHEMA, recordValue, kafkaOffset, 0L, TimestampType.NO_TIMESTAMP_TYPE, null);
 
-    Exception thrownException = assertThrows(DataException.class, () -> writeRecordWithKeysAndHeaders(faultyRecord));
-    String expectedMessage = String.format("Headers cannot be null for SinkRecord: %s", faultyRecord);
-    assertEquals(expectedMessage, thrownException.getMessage());
+    String exceptionMessage = String.format("Headers cannot be null for SinkRecord: %s", faultyRecord.toString());
+    testExceptionReportedToDLQ(faultyRecord, DataException.class, exceptionMessage, false);
+    tearDown(); // clear mock S3 port for follow up test
+    // test with faulty being first in batch
+    testExceptionReportedToDLQ(faultyRecord, DataException.class, exceptionMessage, true);
+  }
+
+  @Test
+  public void testSchemaProjectionExceptionReported() throws Exception {
+    String recordValue = "1";
+    int kafkaOffset = 1;
+    SinkRecord faultyRecord = new SinkRecord(TOPIC, PARTITION, Schema.STRING_SCHEMA, "key",
+        null, recordValue, kafkaOffset, 0L, TimestampType.NO_TIMESTAMP_TYPE, sampleHeaders());
+
+    String exceptionMessage = "Switch between schema-based and schema-less data is not supported";
+    testExceptionReportedToDLQ(faultyRecord, SchemaProjectorException.class, exceptionMessage, false);
+  }
+
+  // test DLQ for lower level exception coming from AvroData
+  @Test
+  public void testDataExceptionReportedIncorrectSchema() throws Exception {
+    Schema sampleValueSchema = createSchema();
+    int kafkaOffset = 1;
+    SinkRecord faultyRecord = new SinkRecord(TOPIC, PARTITION, Schema.STRING_SCHEMA, "key",
+        sampleValueSchema, "1", kafkaOffset, 0L, TimestampType.NO_TIMESTAMP_TYPE, sampleHeaders());
+
+    String exceptionMessage = "Invalid type for STRUCT: class java.lang.String";
+    testExceptionReportedToDLQ(faultyRecord, DataException.class, exceptionMessage, false);
+    tearDown(); // clear mock S3 port for follow up test
+    // test with faulty being first in batch
+    testExceptionReportedToDLQ(faultyRecord, DataException.class, exceptionMessage, true);
   }
 
+  // test DLQ for lower level exception coming from AvroData
   @Test
-  public void testRecordKeysAnsHeadersWritten() throws Exception {
-    writeRecordWithKeysAndHeaders(null);
+  public void testDataExceptionReportedNullValueForNonOptionalSchema() throws Exception {
+    SinkRecord faultyRecord = new SinkRecord(TOPIC, PARTITION, Schema.STRING_SCHEMA, "key",
+        Schema.STRING_SCHEMA, null, 1, 0L, TimestampType.NO_TIMESTAMP_TYPE, sampleHeaders());
```

Comment:
This schema selection is incorrect. 

The schema differs from the other records, and will cause a rotation due to schema compatibility checks. But I don't think this is what you want to check here. See if the suggestion below makes sense: 

```suggestion
        createSchema(), null, 1, 0L, TimestampType.NO_TIMESTAMP_TYPE, sampleHeaders());
```